esmvalcore/local.py

```diff
@@ -547,11 +494,131 @@ def get_multiproduct_filename(attributes, preproc_dir):
+
+    def filename(file):
+        return file.name
+
+    def version(file):
+        return file.facets.get('version', '')
+
+    result = []
+    for _, group in itertools.groupby(sorted(files, key=filename),
+                                      key=filename):
+        duplicates = sorted(group, key=version)
+        latest = duplicates[-1]
+        result.append(latest)
+    return result
+
+
+def find_files(
+    *,
+    debug: bool = False,
+    **facets: FacetValue,
+) -> Union[list[LocalFile], tuple[list[LocalFile], list[Path]]]:
+    """Find files on the local filesystem.
+
+    The directories that are searched for files are defined in
+    :data:`esmvalcore.config.CFG` under the ``'rootpath'`` key using the
+    directory structure defined under the ``'drs'`` key.
+    If ``esmvalcore.config.CFG['rootpath']`` contains a key that matches the
+    value of the ``project`` facet, those paths will be used. If there is no
+    project specific key, the directories in
+    ``esmvalcore.config.CFG['rootpath']['default']`` will be searched.
+
+    See :ref:`findingdata` for extensive instructions on configuring ESMValCore
+    so it can find files locally.
+
+    Parameters
+    ----------
+    debug
+        When debug is set to :obj:`True`, the function will return a tuple
+        with the first element containing the files that were found
+        and the second element containing the globs patterns that
+        were used to search for files.
+    **facets
+        Facets used to search for files. An ``'*'`` can be used to match
+        any value. By default, only the latest version of a file will
+        be returned. To select all versions use ``version='*'``. It is also
+        possible to specify multiple values for a facet, e.g.
+        ``exp=['historical', 'ssp585']`` will match any file that belongs
+        to either the historical or ssp585 experiment.
+
+    Examples
+    --------
+    Search for files containing surface air temperature from any CMIP6 model
+    for the historical experiment:
+
+    >>> esmvalcore.local.find_files(
+    ...     project='CMIP6',
+    ...     activity='CMIP',
+    ...     mip='Amon',
+    ...     short_name='tas',
+    ...     exp='historical',
+    ...     dataset='*',
+    ...     ensemble='*',
+    ...     grid='*',
+    ...     institute='*',
+    ... )  # doctest: +SKIP
+    [LocalFile('/home/bandela/climate_data/CMIP6/CMIP/BCC/BCC-ESM1/historical/r1i1p1f1/Amon/tas/gn/v20181214/tas_Amon_BCC-ESM1_historical_r1i1p1f1_gn_185001-201412.nc')]
+
+    Returns
+    -------
+    list[LocalFile]
+        The files that were found.
+    """
+    filenames, globs = _get_input_filelist(facets)
+    drs = _select_drs('input_dir', facets['project'])
+    if isinstance(drs, list):
+        # Not sure how to handle a list of DRSs
+        drs = ''
+    files = []
+    for filename in filenames:
+        file = LocalFile(filename)
+        file.facets.update(_path2facets(file, drs))
```

Comment:
I added a note to the documentation to warn users about this limitation.

Reply:
What does that mean in practice? Using wildcards for these facets will raise an error? Or will it work but the facet is not updated in `metadata.yml` and remains `*`?

Reply:
Using wildcards in the recipe will be supported once we have all of #1609 merged. This pull request only adds the possibility to use the `esmvalcore.local.find_files` function to search for files using wildcards. When it is not possible to obtain the facets from the directory names, `LocalFile.facets` will be an empty dict, that's all, no errors.

Reply:
> I guess LocalFile.facets will be used to update facets for metadata.yml if wildcards are used (if not, please ignore).

As a starting point, but there are some intermediate steps that make sure no wildcards end up in metadata.yml. Would it be OK with you to postpone this discussion until we get to a pull request that actually implements wildcards in the recipe? Or take this discussion fowards in #1609?

Reply:
Sure, fine for me, just wanted to make you aware of potential problems with some projects :+1:

Reply:
Thanks, that's much appreciated! I just find it easier to discuss those when they occur in the code that is under review because otherwise, the discussion becomes hypothetical and there is a lot of room for misunderstanding.